include/seqan3/alignment/scoring/scoring_scheme_concept.hpp

```diff
@@ -0,0 +1,97 @@
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+// DAMAGE.
+//
+// ============================================================================
+
+/*!\file
+ * \brief Provides seqan3::scoring_scheme_concept.
+ * \author Hannes Hauswedell <hannes.hauswedell AT fu-berlin.de>
+ */
+
+#pragma once
+
+#include <seqan3/alphabet/concept.hpp>
+
+namespace seqan3
+{
+
+/*!\interface seqan3::scoring_scheme_concept <>
+ * \brief A concept that requires that type be able to score two letters.
+ * \tparam t            The type the concept check is performed on (the putative scoring scheme).
+ * \tparam alphabet_t   The type of the first letter that you wish to score; must model seqan3::alphabet_concept.
+ * \tparam alphabet2_t  The type of the second letter that you wish to score; must model seqan3::alphabet_concept;
+ *                      defaults to `alphabet_t`.
+ * \ingroup scoring
+ *
+ * \details
+ *
+ * This concept makes no assumptions about configurability or assignability of the scoring scheme, only the
+ * ability to score the two letters is required.
+ *
+ */
+/*!\name Requirements for seqan3::scoring_scheme_concept
+ * \brief You can expect these members on all types that implement seqan3::scoring_scheme_concept.
+ * \memberof seqan3::scoring_scheme_concept
+ * \{
+ */
+/*!\typedef     typedef IMPLEMENTATION_DEFINED score_type;
+ * \brief       The type returned by seqan3::scoring_scheme_concept::score(), usually a seqan3::arithmetic_concept.
+ * \memberof seqan3::scoring_scheme_concept
+ *
+ * \details
+ * \attention This is a concept requirement, not an actual typedef (however types satisfying this concept
+ * will provide an implementation).
+ */
+/*!\fn          score_type score(alph1_t const alph1, alph2_t const alph2);
+ * \brief       Compute the score of two letters
+ * \param alph1 First letter.
+ * \param alph2 Second letter.
+ * \memberof seqan3::scoring_scheme_concept
+ *
+ * \details
+ * \attention This is a concept requirement, not an actual function (however types satisfying this concept
+ * will provide an implementation).
+ */
+//!\}
+//!\cond
+template <typename t, alphabet_concept alphabet_t, alphabet_concept alphabet2_t = alphabet_t>
+concept bool scoring_scheme_concept = requires (t scheme,
+                                                alphabet_t const alph1,
+                                                alphabet2_t const alph2)
+{
+    { scheme.score(alph1, alph2) } -> common_reference_concept<typename std::remove_reference_t<t>::score_type>;
```

Comment:
In my opinion, if this is a concept thought for score classes to be customisation points, we should offer them as global free functions.

Reply:
but the scoring schemes are very specific to our algorithms. People might want to create their own, but I can't think of a place where they would want to integrate this with another library or need to wrap another library's type...

Reply:
On the contrary, people might actually have their own library, want to use our algorithms, but have different scoring schemes. Like if you would need a PSSM, which we don't offer (yet).  They either need to write a wrapper around their scheme or just need to provide an overload for `scrore(scheme, alph1, alph2)`.